src/gen9_avc_encoder.c

```diff
@@ -3596,6 +3662,456 @@ gen9_avc_set_curbe_mbenc(VADriverContextP ctx,
+
+    if (generic_state->frame_type == SLICE_TYPE_I) {
+        cmd->dw0.skip_mode_enable = 0;
+        cmd->dw37.skip_mode_enable = 0;
+        cmd->dw36.hme_combine_overlap = 0;
+        cmd->dw36.check_all_fractional_enable = 0;
+        cmd->dw47.intra_cost_sf = 16;/* not used, but recommended to set 16 by kernel team */
+        cmd->dw34.enable_direct_bias_adjustment = 0;
+        cmd->dw34.enable_global_motion_bias_adjustment = 0;
+
+    } else if (generic_state->frame_type == SLICE_TYPE_P) {
+        cmd->dw1.max_num_mvs = i965_avc_get_max_mv_per_2mb(avc_state->seq_param->level_idc) / 2;
+        cmd->dw3.bme_disable_fbr = 1;
+        cmd->dw5.ref_width = cmd->dw39.ref_width = ref_width;
+        cmd->dw5.ref_height = cmd->dw39.ref_height = ref_height;
+        cmd->dw7.non_skip_zmv_added = 1;
+        cmd->dw7.non_skip_mode_added = 1;
+        cmd->dw7.skip_center_mask = 1;
+
+        cmd->dw47.intra_cost_sf =
+            (avc_state->adaptive_intra_scaling_enable) ?
+            gen9_avc_adaptive_intra_scaling_factor[preset] :
+            gen9_avc_intra_scaling_factor[preset];
+
+        cmd->dw47.max_vmv_r =
+            i965_avc_get_max_mv_len(avc_state->seq_param->level_idc) * 4;//frame only
+
+        cmd->dw36.hme_combine_overlap = 1;
+        cmd->dw36.num_ref_idx_l0_minus_one = (avc_state->multi_pre_enable) ? slice_param->num_ref_idx_l0_active_minus1 : 0;
+        cmd->dw36.check_all_fractional_enable = fei_param->repartition_check_enable;
+        cmd->dw34.enable_direct_bias_adjustment = 0;
+        cmd->dw34.enable_global_motion_bias_adjustment = avc_state->global_motion_bias_adjustment_enable;
+        if (avc_state->global_motion_bias_adjustment_enable)
+            cmd->dw59.hme_mv_cost_scaling_factor = avc_state->hme_mv_cost_scaling_factor;
+
+        cmd->dw64.num_mv_predictors_l0 = fei_param->num_mv_predictors_l0;
+
+    } else { /* B slice */
+
+        cmd->dw1.max_num_mvs = i965_avc_get_max_mv_per_2mb(avc_state->seq_param->level_idc) / 2;
+        cmd->dw1.bi_Weight = avc_state->bi_weight;
+        cmd->dw3.search_ctrl = 7;
+        cmd->dw3.skip_type = 1;
+        cmd->dw5.ref_width = cmd->dw39.ref_width = ref_width;
+        cmd->dw5.ref_height = cmd->dw39.ref_height = ref_height;
+        cmd->dw7.skip_center_mask = 0xff;
+
+        cmd->dw47.intra_cost_sf = avc_state->adaptive_intra_scaling_enable ?
+                                  gen9_avc_adaptive_intra_scaling_factor[qp] :
+                                  gen9_avc_intra_scaling_factor[qp];
+
+        cmd->dw47.max_vmv_r =
+            i965_avc_get_max_mv_len(avc_state->seq_param->level_idc) * 4;//frame only
+
+        cmd->dw36.hme_combine_overlap = 1;
+
+        //check is_fwd_frame_short_term_ref
+        surface_id = slice_param->RefPicList1[0].picture_id;
+        obj_surface = SURFACE(surface_id);
+        if (!obj_surface) {
+            WARN_ONCE("Invalid backward reference frame\n");
+            return;
```

Comment:
Somehow I missed this comment :).
Good catch! But need a similar fix in a number of places which includes non-fei use cases too !
I will fix it for FEI use case in this patch.